nvme.pyx

```diff
@@ -53,3 +113,206 @@ def get_nsid(path):
+        cdef uint64_t icrkey = crkey
+        cdef uint8_t irrela = rrela
+        cdef uint8_t irtype = rtype
+
+        data[0] = defs.htole64(icrkey)
+
+        memset(&pt, 0, sizeof(pt))
+        pt.cmd.opc = NVME_OPC_RESERVATION_RELEASE
+        pt.cmd.cdw10 = defs.htole32((irrela & 7) | (irtype << 8))
+        pt.buf = &data
+        pt.len = sizeof(data)
+        pt.is_read = 0
+
+        with nogil:
+            res = ioctl(self.fd, NVME_PASSTHROUGH_CMD, &pt)
+            if res == -1:
+                raise OSError('Release request failed')
+
+            sc = NVME_STATUS_GET_SC(pt.cpl.status)
+            st = NVME_STATUS_GET_SCT(pt.cpl.status)
+            if sc != 0 or st != 0:
+                raise OSError('Release request returned error')
+
+        return True
+
+    def resvreport(self):
+        cdef nvme_pt_command pt
+        cdef nvme_resv_status *s
+        cdef uint8_t data[4096]
+        cdef unsigned int i, n
+        cdef bint eds = False
+
+        bzero(data, sizeof(data))
+        memset(&pt, 0, sizeof(pt))
+        pt.cmd.opc = NVME_OPC_RESERVATION_REPORT
+        pt.cmd.cdw10 = defs.htole32(sizeof(data) // 4 - 1)
+        pt.cmd.cdw11 = defs.htole32(sizeof(eds))
+        pt.buf = &data
+        pt.len = sizeof(data)
+        pt.is_read = 1
+
+        with nogil:
+            res = ioctl(self.fd, NVME_PASSTHROUGH_CMD, &pt)
+            if res < 0:
+                raise OSError('Report request failed')
+
+            sc = NVME_STATUS_GET_SC(pt.cpl.status)
+            st = NVME_STATUS_GET_SCT(pt.cpl.status)
+            if sc != 0 or st != 0:
+                raise OSError('Report request returned error')
+
+        nvme_resv_status_swapbytes(<void *>data, sizeof(data))
+
+        info = {}
+        s = <nvme_resv_status *>data
+        n = (s.regctl[1] << 8) | s.regctl[0]
+        info['generation'] = s.gen
+        info['scopetype'] = s.rtype
+        info['number_of_registered_controllers'] = n
+        info['persist_through_power_loss_state'] = s.ptpls
+        info['controllers'] = []
+
+        n = min(n, (sizeof(data) - sizeof(s)) // sizeof(s.ctrlr[0]))
+
+        controller = {}
+        for i in range(n):
+            controller['controller_id'] = s.ctrlr[i].ctrlr_id
+            controller['resv_status'] = s.ctrlr[i].rcsts
+            controller['host_id'] = s.ctrlr[i].hostid
+            controller['key'] = s.ctrlr[i].rkey
+            info['controllers'].append(controller.copy())
+
+        return info
+
+    def read_reservation(self):
+        """
+        This function returns:
+            1. the number of keys that
+                have been put on the disk (generation)
+            2. the type of reservation that
+                is being held (scopetype)
+            3. the reservation key that reserves
+                the disk (reservation)
+        """
+
+        data = self.resvreport()
+
+        for i in data['controllers']:
+            if i.get('resv_status') == 1:
```

Comment:
Is `resv_status` ever going to not exist ? I think no, in this case we should access it directly so it fails we know something is off.